etcdserver/v3_server.go

```diff
@@ -805,5 +807,88 @@ func (s *EtcdServer) AuthInfoFromCtx(ctx context.Context) (*auth.AuthInfo, error
 }
 
 func (s *EtcdServer) Downgrade(ctx context.Context, r *pb.DowngradeRequest) (*pb.DowngradeResponse, error) {
-	return nil, nil
+	switch r.Action {
+	case pb.DowngradeRequest_VALIDATE:
+		return s.downgradeValidate(ctx, r.Version)
+	case pb.DowngradeRequest_ENABLE:
+		return s.downgradeEnable(ctx, r)
+	case pb.DowngradeRequest_CANCEL:
+		return s.downgradeCancel(ctx)
+	default:
+		return nil, ErrUnknownDowngradeAction
+	}
+}
+
+func (s *EtcdServer) downgradeValidate(ctx context.Context, v string) (*pb.DowngradeResponse, error) {
+	resp := &pb.DowngradeResponse{}
+
+	targetVersion, err := changeToClusterVersion(v)
+	if err != nil {
+		return nil, err
+	}
+
+	// do linearized read to avoid using stale downgrade information
+	err = s.linearizableReadNotify(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	cv := s.ClusterVersion()
+	if cv == nil {
+		return nil, ErrClusterVersionUnavailable
+	}
+	resp.Version = cv.String()
+
+	allowedTargetVersion := semver.Version{Major: cv.Major, Minor: cv.Minor - 1}
```

Comment:
In the future if cluster version is 4.0, does this work?

Reply:
If the cluster version is 4.0, I think the downgrade target should be 3.x which is the latest v3 minor version according to downgrade policy. Since we cannot predict what the 3.x could be, I don't handle this case. I encapsulate this line into a function called `allowedDowngradeVersion` and add a todo there. It deserve an update before we bump to v4.0. WDYT?